src/main/java/duke/main/Duke.java

```diff
@@ -0,0 +1,199 @@
+         * It also handles duke.exceptions and waits for user to say bye.
+         */
+        protected void run() {
+
+            Scanner scanner = new Scanner(System.in);
+            String input = "";
+            while (!input.equals("bye")) {
+                try {
+                    input = scanner.nextLine();
+                    String command = this.parser.getCommand(input);
+                    switch (command) {
+                        case "list":
+                            taskManager.list();
+                            break;
+                        case "mark":
+                        case "unmark":
+                            handleMarking(input, taskManager);
+                            break;
+                        case "todo":
+                            handleTodo(input, taskManager);
+                            break;
+                        case "deadline":
+                            handleDeadline(input, taskManager);
+                            break;
+                        case "event":
+                            handleEvent(input, taskManager);
+                            break;
+                        case "delete":
+                            handleDelete(input, taskManager);
+                            break;
+                        default:
+                            throw new UnknownCommandException("I may be the GOAT but I don't know what that means.");
+                    }
+                } catch (DukeException e) {
+                    ui.displayError(e);
+                }
+            }
+            ui.exit();
+        }
+
+
+        /**
+         * returns the first word of the input string. Helps to determine which function to call.
+         * @param input
+         * @return command string
+         */
+
+        /**
+         * This function handles the logic for when a user wants to mark a task as done or undone.
+         * @param input
+         * @param taskManager
+         * @throws InvalidArgumentException
+         */
+        private void handleMarking(String input, TaskManager taskManager) throws DukeException {
+            String[] words = input.split(" ");
+            try {
+                int index = Integer.parseInt(words[1]);
+                if (words[0].equals("mark")) {
+                    taskManager.mark(index);
+                } else if (words[0].equals("unmark")) {
+                    taskManager.unmark(index);
+                }
+                updateStorage();
+            } catch (NumberFormatException e) {
+                throw new InvalidArgumentException("Please enter a numerical index!");
+            } catch (StorageException e) {
+                throw new StorageException("duke.main.Storage Error");
+            }
+        }
+
+        private void updateStorage() throws StorageException {
+            this.storage.saveData(taskManager);
+        }
```

Comment:
I think it might be better if all the command handling are contained inside the TaskManager class.